File: _708_InsertIntoASortedCircularLinkedList.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next
"""

class Solution: # Official approach
    def insert(self, head: 'Optional[Node]', insertVal: int) -> 'Node':
        if head == None:
            node = Node(insertVal, None)
            node.next = node
            return node
        
        prev, cur = head, head.next
        toInsert = False
        while True:# how to end?
            if prev.val <= insertVal <= cur.val:#Among the largest and smallest
                toInsert = True
            elif prev.val > cur.val:# It is the largest or smallest
                if insertVal < cur.val or insertVal > prev.val:
                    toInsert = True
            if toInsert:
                prev.next = Node(insertVal, cur)
                return head
            prev, cur = cur, cur.next
            # Only one node
            if prev == head:
                break
        
        prev.next = Node(insertVal, cur)
        return head

class Solution:
    def insert(self, head: 'Optional[Node]', insertVal: int) -> 'Node':
        if head == None:
            node = Node(insertVal, None)
            node.next = node
            return node
        
        prev, cur = head, head.next
        while True:
            if prev.val <= insertVal <= cur.val:
                node = Node(insertVal, cur)
                prev.next = node
                return head
            elif prev.val > cur.val:
                if insertVal < cur.val or insertVal > prev.val:
                    node = Node(insertVal, cur)
                    prev.next = node
                    return head
        # It's possible that sorted list can contain some duplicate values
            prev, cur = cur, cur.next#One loop over, couldn't insert
            if prev == head:
                break

        prev.next = Node(insertVal, cur)
        return head

File: _708_InsertIntoASortedCircularLinkedList.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _next) {
        val = _val;
        next = _next;
    }
};
*/

class Solution {
    public Node insert(Node head, int insertVal) {
        if (head == null) {
            Node node = new Node(insertVal);
            node.next = node;
            return node;
        }
        Node prev = head, curr = head.next;
        Node node = new Node(insertVal);
        while (true) {
            // Case 1. It's descending order, larger than prev,
            // smaller than curr
            if (insertVal >= prev.val &&
                insertVal <= curr.val) {
                prev.next = node;
                node.next = curr;
                return head;
            }
            // Case 2. Larger than all, or smaller than all, inserted
            // into between largest and smallest
            if (prev.val > curr.val) {
                if (insertVal <= curr.val || insertVal >= prev.val) {
                    prev.next = node;
                    node.next = curr;
                    return head;
                }
            }
            prev = curr;
            curr = curr.next;
            // Case 3. Loop all over but not in case 1 or 2.  
            // Therefore there're duplicate elements filled 
            // in the ring. Break the loop and inserted anywhere.
            if (prev == head) break;
        }
        node.next = curr;
        prev.next = node;
        return head;
    }
}